src/Orleans.Runtime/Silo/Silo.cs

```diff
@@ -702,106 +720,102 @@ private void Terminate(bool gracefully)
                     logger.Info(ErrorCode.WaitingForSiloStop, "Waiting {0} for termination to complete", pause);
                     Thread.Sleep(pause);
                 }
-                return;
+                return this.siloTerminatedTask.Task;
             }
+            return this.siloLifecycle.OnStop(cancellationToken);
+        }
 
+        private Task OnRuntimeServicesStop(CancellationToken ct)
+        {
+            string operation = gracefully ? "Shutdown()" : "Stop()";
             try
             {
-                try
-                {
-                    if (gracefully)
-                    {
-                        logger.Info(ErrorCode.SiloShuttingDown, "Silo starting to Shutdown()");
-                        // 1: Write "ShutDown" state in the table + broadcast gossip msgs to re-read the table to everyone
-                        scheduler.QueueTask(this.membershipOracle.ShutDown, this.membershipOracleContext)
-                            .WaitWithThrow(stopTimeout);
-                    }
-                    else
-                    {
-                        logger.Info(ErrorCode.SiloStopping, "Silo starting to Stop()");
-                        // 1: Write "Stopping" state in the table + broadcast gossip msgs to re-read the table to everyone
-                        scheduler.QueueTask(this.membershipOracle.Stop, this.membershipOracleContext)
-                            .WaitWithThrow(stopTimeout);
-                    }
-                }
-                catch (Exception exc)
+                if (this.gracefully)
                 {
-                    logger.Error(ErrorCode.SiloFailedToStopMembership, String.Format("Failed to {0} membership oracle. About to FastKill this silo.", operation), exc);
-                    return; // will go to finally
+                    logger.Info(ErrorCode.SiloShuttingDown, "Silo starting to Shutdown()");
+                    // 1: Write "ShutDown" state in the table + broadcast gossip msgs to re-read the table to everyone
+                    scheduler.QueueTask(this.membershipOracle.ShutDown, this.membershipOracleContext)
+                        .WaitWithThrow(stopTimeout);
```

Comment:
Use `await xxx.WithTimeout` everywhere and turn the method to be fully async. Otherwise the `LifecycleObservable` will not parallelize as expected.
The same with the `OnRuntimeInitializeStop` method.

Reply:
I'll change what I can, but a lot of our startup logic is synchronous and it's out of scope for this PR to significantly refactor all of that.  For instance, there are many cases where async actions are wrapped in SafeExecute, which assumes synchronous actions.

Reply:
Sure, address the trivial ones. As long as the first currently blocking operation just yields, then we already achieve parallelism, we can address others later